File: server/security/validators.py
# server/security/validators.py

import re


class ValidationError(Exception):
    """
    Raised when user input validation fails.
    """
    pass


# -------------------------------------------------
# BASIC SANITIZATION
# -------------------------------------------------
def sanitize_text(value: str) -> str:
    """
    Trim whitespace and remove control characters.
    """
    if not isinstance(value, str):
        raise ValidationError("Invalid input type")

    value = value.strip()

    # Remove ASCII control characters
    value = re.sub(r"[\x00-\x1f\x7f]", "", value)

    return value


# -------------------------------------------------
# NAME VALIDATION
# -------------------------------------------------
def validate_name(name: str) -> str:
    """
    Validate student name.
    Allowed:
        - Letters (A–Z, a–z)
        - Spaces
        - Dot (.)
    """
    name = sanitize_text(name)

    if not (2 <= len(name) <= 100):
        raise ValidationError("Name length must be between 2 and 100 characters")

    if not re.fullmatch(r"[A-Za-z.\s]+", name):
        raise ValidationError("Name contains invalid characters")

    return name


# -------------------------------------------------
# ROLL NUMBER VALIDATION
# -------------------------------------------------
def validate_roll(roll: str) -> str:
    """
    Validate roll number.
    Allowed:
        - Alphanumeric only
        - No spaces
        - No special characters
    """
    roll = sanitize_text(roll)

    if not (1 <= len(roll) <= 30):
        raise ValidationError("Roll number length must be between 1 and 30 characters")

    if not re.fullmatch(r"[A-Za-z0-9\-]+", roll):
        raise ValidationError("Roll number must be alphanumeric only")

    return roll.upper()


# -------------------------------------------------
# SESSION CODE VALIDATION
# -------------------------------------------------
def validate_session_code(code: str, expected_code: str) -> None:
    """
    Validate session code entered by student.
    The code is NOT stored in DB.
    """
    code = sanitize_text(code)

    if code != expected_code:
        raise ValidationError("Invalid session code")

File: internal/security/validators.go
// internal/security/validatior.go

package security

import (
	"fmt"
	"regexp"
	"strings"
)

type ValidationError struct{ msg string }

func (e *ValidationError) Error() string { return e.msg }

func newValidationErr(format string, a ...any) error {
	return &ValidationError{msg: fmt.Sprintf(format, a...)}
}

var (
	nameRe = regexp.MustCompile(`^[A-Za-z.\s]+$`)
	rollRe = regexp.MustCompile(`^[A-Za-z0-9\-]+$`)
	ctrlRe = regexp.MustCompile(`[\x00-\x1f\x7f]`)
)

func SanitizeText(value string) string {
	value = strings.TrimSpace(value)
	return ctrlRe.ReplaceAllString(value, "")
}

func ValidateName(name string, maxLen int) (string, error) {
	name = SanitizeText(name)
	if len(name) < 2 || len(name) > maxLen {
		return "", newValidationErr("name length must be between 2 and %d characters", maxLen)
	}
	if !nameRe.MatchString(name) {
		return "", newValidationErr("name contains invalid characters")
	}
	return name, nil
}

func ValidateRoll(roll string, maxLen int) (string, error) {
	roll = SanitizeText(roll)
	if len(roll) < 1 || len(roll) > maxLen {
		return "", newValidationErr("roll number length must be between 1 and %d characters", maxLen)
	}
	if !rollRe.MatchString(roll) {
		return "", newValidationErr("roll number must be alphanumeric only")
	}
	return strings.ToUpper(roll), nil
}

func ValidateSessionCode(code, expected string) error {
	code = SanitizeText(code)
	if code != expected {
		return newValidationErr("invalid session code")
	}
	return nil
}
